plasmapy/utils/decorators/tests/test_validators.py

```diff
@@ -0,0 +1,490 @@
+            {'input': {'args': (5. * u.cm, 'arg'),
+                       'validations': {**default_validations,
+                                       'units': [u.km]}},
+             'output': (5. * u.cm).to(u.km)},
+
+            # return value is None and not allowed
+            {'input': {'args': (None, 'validations_on_return'),
+                       'validations': {**default_validations,
+                                       'units': [u.cm],
+                                       'none_shall_pass': False}},
+             'raises': ValueError},
+
+            # 'pass_equivalent_units' is True and unit conversion is not performed
+            {'input': {'args': (5 * u.cm, 'arg'),
+                       'validations': {**default_validations,
+                                       'units': [u.km],
+                                       'pass_equivalent_units': True}},
+             'output': 5 * u.cm},
+        ]
+
+        # setup wrapped function
+        vq = ValidateQuantities()
+        vq.f = self.foo
+
+        # perform tests
+        for ii, case in enumerate(_cases):
+            arg, arg_name = case['input']['args']
+            validations = case['input']['validations']
+
+            if 'warns' in case:
+                with pytest.warns(case['warns']):
+                    warnings.simplefilter("always",
+                                          category=ImplicitUnitConversionWarning)
+                    _result = vq._validate_quantity(arg, arg_name, validations)
+            elif 'raises' in case:
+                with pytest.raises(case['raises']):
+                    vq._validate_quantity(arg, arg_name, validations)
+                continue
+            else:
+                _result = vq._validate_quantity(arg, arg_name, validations)
+
+            assert _result == case['output']
+
+        # method calls `_check_unit_core` and `_check_value`
+        case = {'input': (5. * u.cm, u.cm, {**default_validations, 'units': [u.cm]}),
+                'output': 5. * u.cm}
+        with mock.patch.object(CheckUnits,
+                               '_check_unit_core',
+                               return_value=(5*u.cm, u.cm, None, None)) \
+                as mock_cu_checks, \
+                mock.patch.object(CheckValues,
+                                  '_check_value',
+                                  return_value=None) as mock_cv_checks:
+
+            args = case['input'][0:2]
+            validations = case['input'][2]
+
+            vq = ValidateQuantities(**validations)
+            vq.f = self.foo
+
+            assert vq._validate_quantity(*args, validations) == case['output']
+            assert mock_cu_checks.called
+            assert mock_cv_checks.called
+
+    def test_vq_preserves_signature(self):
+        """Test `ValidateQuantities` preserves signature of wrapped function."""
+        # I'd like to directly dest the @preserve_signature is used (??)
```

Comment:
Inconsequential, I know, but, typo fix:
```suggestion
        # I'd like to directly test the @preserve_signature is used (??)
```

I'm not sure how this goal could be achieved without profiling which functions are being called... and I think that might be more trouble than it's worth.